File: src/foam/nanos/notification/email/SMTPEmailService.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.nanos.notification.email',
  name: 'SMTPEmailService',

  implements: [
    'foam.nanos.notification.email.EmailService'
  ],

  documentation: 'Implementation of Email Service using SMTP',

  imports: [
    'threadPool?'
  ],

  javaImports: [
    'foam.nanos.notification.email.SMTPConfig',
    'foam.nanos.logger.Logger',
    'foam.nanos.om.OMLogger',
    'static foam.mlang.MLang.EQ',
    'java.util.Date',
    'java.util.Properties',
    'javax.mail.*',
    'javax.mail.internet.InternetAddress',
    'javax.mail.internet.MimeMessage',
    'org.apache.commons.lang3.StringUtils',
  ],

  axioms: [
    {
      name: 'javaExtras',
      buildJavaClass: function(cls) {
        cls.extras.push(foam.java.Code.create({
          data:
            `private class SMTPAuthenticator extends javax.mail.Authenticator {
              protected String username_;
              protected String password_;

              public SMTPAuthenticator(String username, String password) {
                this.username_ = username;
                this.password_ = password;
              }

              @Override
              protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication(this.username_, this.password_);
              }
            }
            `
        }));
      }
    }
  ],

  properties: [
    {
      name: 'session_',
      javaType: 'Session',
      class: 'Object',
      javaFactory:
      `
        SMTPConfig smtpConfig = (SMTPConfig) getX().get("SMTPConfig");
        Properties props = new Properties();
        props.setProperty("mail.smtp.auth", smtpConfig.getAuthenticate() ? "true" : "false");
        props.setProperty("mail.smtp.starttls.enable", smtpConfig.getStarttls() ? "true" : "false");
        props.setProperty("mail.smtp.host", smtpConfig.getHost());
        props.setProperty("mail.smtp.port", smtpConfig.getPort());
        if ( smtpConfig.getAuthenticate() ) {
          return Session.getInstance(props, new SMTPAuthenticator(smtpConfig.getUsername(), smtpConfig.getPassword()));
        }
        return Session.getInstance(props);
      `
    },
    {
      class: 'Object',
      javaType: 'Transport',
      name: 'transport_',
      javaFactory:
      `
        Logger logger = (Logger) getX().get("logger");
        OMLogger omLogger = (OMLogger) getX().get("OMLogger");
        SMTPConfig smtpConfig = (SMTPConfig) getX().get("SMTPConfig");
        Transport transport = null;
        try {
          omLogger.log(this.getClass().getSimpleName(), "transport", "connecting");
          transport = getSession_().getTransport("smtp");
          transport.connect(smtpConfig.getUsername(), smtpConfig.getPassword());
          logger.info("SMTPEmailService connected.");
          omLogger.log(this.getClass().getSimpleName(), "transport", "connected");
        } catch ( Exception e ) {
          logger.error("Transport failed to initialize: " + e);
        }
        return transport;
      `
    },
  ],

  methods: [
    {
      name: 'createMimeMessage',
      javaType: 'MimeMessage',
      documentation: `Create a MimeMessage from the passed EmailMessage`,
      args: [
        {
          name: 'emailMessage',
          type: 'foam.nanos.notification.email.EmailMessage'
        }
      ],
      javaCode:
      `
        Logger logger = (Logger) getX().get("logger");
        try {
          MimeMessage message = new MimeMessage(getSession_());

          // the From Property is mainly to hide our smtp user credetials.
          if ( emailMessage.isPropertySet("from") ) {
            if ( emailMessage.isPropertySet("displayName") ) {
              message.setFrom( new InternetAddress(emailMessage.getFrom(), emailMessage.getDisplayName()) );
            } else {
              message.setFrom(new InternetAddress(emailMessage.getFrom()));
            }
          } else if ( emailMessage.isPropertySet("replyTo") ) {
              if ( emailMessage.isPropertySet("displayName") ) {
                message.setFrom( new InternetAddress(emailMessage.getReplyTo(), emailMessage.getDisplayName()) );
              } else {
                message.setFrom(new InternetAddress(emailMessage.getReplyTo()));
              }
          }

          if ( emailMessage.isPropertySet("replyTo") )
            message.setReplyTo(InternetAddress.parse(emailMessage.getReplyTo()));

          if ( emailMessage.isPropertySet("subject") )
            message.setSubject(emailMessage.getSubject());

          if ( emailMessage.isPropertySet("body") )
            message.setContent(emailMessage.getBody(), "text/html; charset=utf-8");

          if ( emailMessage.isPropertySet("to") ) {
            if ( emailMessage.getTo().length == 1 ) {
              message.setRecipient(Message.RecipientType.TO, new InternetAddress((emailMessage.getTo())[0], false));
            } else {
              message.setRecipients(Message.RecipientType.TO, InternetAddress.parse(StringUtils.join(emailMessage.getTo(), ",")));
            }
          }

          if ( emailMessage.isPropertySet("cc") ) {
            if ( emailMessage.getCc().length == 1 ) {
              message.setRecipient(Message.RecipientType.CC, new InternetAddress((emailMessage.getCc())[0], false));
            } else {
              message.setRecipients(Message.RecipientType.CC, InternetAddress.parse(StringUtils.join(emailMessage.getCc(), ",")));
            }
          }

          if ( emailMessage.isPropertySet("bcc") ) {
            if ( emailMessage.getBcc().length == 1 ) {
              message.setRecipient(Message.RecipientType.BCC, new InternetAddress((emailMessage.getBcc())[0], false));
            } else {
              message.setRecipients(Message.RecipientType.BCC, InternetAddress.parse(StringUtils.join(emailMessage.getBcc(), ",")));
            }
          }

          message.setSentDate(new Date());
          logger.info("SMTPEmailService Created MimeMessage.");
          return message;
        } catch (Throwable t) {
          logger.error("SMTPEmailService failed to created MimeMessage. " + t);
          return null;
        }
      `
    },
    {
      name: 'sendEmail',
      javaCode: `
        Logger logger = (Logger) getX().get("logger");
        OMLogger omLogger = (OMLogger) getX().get("OMLogger");

        emailMessage = (EmailMessage) emailMessage.fclone();
        if ( emailMessage.getStatus() == Status.FAILED ) {
          // ignore
          logger.debug("Email not sent, already FAILED.", emailMessage.getId());
          return emailMessage;
        }

        MimeMessage message = createMimeMessage(emailMessage);
        if ( message == null ) {
          return emailMessage;
        }
        try {
          getTransport_().send(message);
          emailMessage.setStatus(Status.SENT);
          logger.debug("SMTPEmailService sent MimeMessage.");
          omLogger.log(this.getClass().getSimpleName(), "message", "sent");
        } catch ( SendFailedException e ) {
          emailMessage.setStatus(Status.FAILED);
          logger.error("SMTPEmailService sending MimeMessage failed. " + e);
        } catch ( MessagingException e ) {
          try {
            getTransport_().close();
          } catch ( Exception e2 ) {
            logger.error("Failed to close transport. " + e2);
          }
          clearTransport_();
          logger.error("SMTPEmailService sending MimeMessage failed. " + e);
        }
        return emailMessage;
      `
    }
  ]
});

File: src/foam/nanos/notification/email/SMTPEmailService.java
package foam.nanos.notification.email;

import foam.core.ContextAgent;
import foam.core.ContextAwareSupport;
import foam.core.X;
import foam.nanos.NanoService;
import foam.nanos.auth.User;
import foam.nanos.pool.FixedThreadPool;
import org.apache.commons.lang3.StringUtils;
import org.jtwig.JtwigModel;
import org.jtwig.JtwigTemplate;
import org.jtwig.environment.EnvironmentConfiguration;
import org.jtwig.environment.EnvironmentConfigurationBuilder;
import org.jtwig.resource.loader.TypedResourceLoader;

import javax.mail.Message;
import javax.mail.PasswordAuthentication;
import javax.mail.Session;
import javax.mail.Transport;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.util.Date;
import java.util.Map;
import java.util.Properties;

public class SMTPEmailService
  extends ContextAwareSupport
  implements EmailService, NanoService
{
  private class SMTPAuthenticator extends javax.mail.Authenticator {
    @Override
    protected PasswordAuthentication getPasswordAuthentication() {
      return new PasswordAuthentication(username_, password_);
    }
  }

  protected Session session_ = null;
  protected EnvironmentConfiguration config_ = null;

  protected String host_;
  protected String port_;
  protected boolean authenticate_;
  protected boolean starttls_;
  protected String username_;
  protected String password_;

  public static class Builder
      extends ContextAwareSupport
  {
    private String host_ = "127.0.0.1";
    private String port_ = "25";
    private boolean authenticate_ = false;
    private boolean starttls_ = false;
    private String username_ = null;
    private String password_ = null;

    public Builder(X x) {
      setX(x);
    }

    public Builder setHost(String host) {
      host_ = host;
      return this;
    }

    public Builder setPort(String port) {
      port_ = port;
      return this;
    }

    public Builder setAuthenticate(boolean authenticate) {
      authenticate_ = authenticate;
      return this;
    }

    public Builder setStartTLS(boolean starttls) {
      starttls_ = starttls;
      return this;
    }

    public Builder setUsername(String username) {
      username_ = username;
      return this;
    }

    public Builder setPassword(String password) {
      this.password_ = password;
      return this;
    }

    public SMTPEmailService build() {
      return new SMTPEmailService(getX(), this);
    }
  }

  protected SMTPEmailService(X x, Builder builder) {
    setX(x);
    host_ = builder.host_;
    port_ = builder.port_;
    authenticate_ = builder.authenticate_;
    starttls_ = builder.starttls_;
    username_ = builder.username_;
    password_ = builder.password_;
  }

  public EnvironmentConfiguration getConfig(String group) {
    if ( config_ == null ) {
      config_ = EnvironmentConfigurationBuilder
          .configuration()
          .resources()
          .resourceLoaders()
          .add(new TypedResourceLoader("dao", new DAOResourceLoader(getX(), group)))
          .and().and()
          .build();
    }
    return config_;
  }

  @Override
  public void start() {
    Properties props = new Properties();
    props.setProperty("mail.smtp.auth", authenticate_ ? "true" : "false");
    props.setProperty("mail.smtp.starttls.enable", starttls_ ? "true" : "false");
    props.setProperty("mail.smtp.host", host_);
    props.setProperty("mail.smtp.port", port_);
    if ( authenticate_ ) {
      session_ = Session.getInstance(props, new SMTPAuthenticator());
    } else {
      session_ = Session.getInstance(props);
    }
  }

  @Override
  public void sendEmail(final EmailMessage emailMessage) {
    FixedThreadPool threadPool = (FixedThreadPool) getX().get("threadPool");
    threadPool.submit(getX(), new ContextAgent() {
      @Override
      public void execute(X x) {
        try {
          MimeMessage message = createMimeMessage(emailMessage);
          if ( message == null ) {
            return;
          }

          // send message
          Transport transport = session_.getTransport("smtp");
          transport.connect();
          transport.sendMessage(message, message.getAllRecipients());
          transport.close();
        } catch (Throwable t) {
          t.printStackTrace();
        }
      }
    });
  }

  @Override
  public void sendEmailFromTemplate(User user, EmailMessage emailMessage, String name, Map<String, Object> templateArgs) {
    String group = user != null ? (String) user.getGroup() : null;
    EmailTemplate emailTemplate = DAOResourceLoader.findTemplate(getX(), name, group);
    if ( emailMessage == null )
      return;

    EnvironmentConfiguration config = getConfig(group);
    JtwigTemplate template = JtwigTemplate.inlineTemplate(emailTemplate.getBody(), config);
    JtwigModel model = JtwigModel.newModel(templateArgs);
    emailMessage.setSubject(emailTemplate.getSubject());
    emailMessage.setBody(template.render(model));
    sendEmail(emailMessage);
  }

  protected MimeMessage createMimeMessage(EmailMessage emailMessage) {
    try {
      MimeMessage message = new MimeMessage(session_);

      // don't send email if no sender
      String from = emailMessage.getFrom();
      if ( from == null || from.isEmpty() )
        return null;
      message.setFrom(new InternetAddress(from));

      // don't send email if no subject
      String subject = emailMessage.getSubject();
      if ( subject == null || subject.isEmpty() )
        return null;
      message.setSubject(subject);

      // don't send email if no body
      String body = emailMessage.getBody();
      if ( body == null || body.isEmpty() )
        return null;
      message.setContent(body, "text/html; charset=utf-8");

      // don't send email if no recipient
      String[] to = emailMessage.getTo();
      if ( to == null || to.length <= 0 )
        return null;

      if ( to.length == 1 ) {
        message.setRecipient(Message.RecipientType.TO, new InternetAddress(to[0], false));
      } else {
        message.setRecipients(Message.RecipientType.TO, InternetAddress.parse(StringUtils.join(to, ",")));
      }

      // send email even if no CC
      String[] cc = emailMessage.getCc();
      if ( cc != null && cc.length == 1 ) {
        message.setRecipient(Message.RecipientType.CC, new InternetAddress(cc[0], false));
      } else if ( cc != null && cc.length > 1 ) {
        message.setRecipients(Message.RecipientType.CC, InternetAddress.parse(StringUtils.join(cc, ",")));
      }

      // send email even if no BCC
      String[] bcc = emailMessage.getBcc();
      if ( bcc != null && bcc.length == 1 ) {
        message.setRecipient(Message.RecipientType.BCC, new InternetAddress(bcc[0], false));
      } else if ( bcc != null && bcc.length > 1 ) {
        message.setRecipients(Message.RecipientType.BCC, InternetAddress.parse(StringUtils.join(bcc, ",")));
      }

      // set date
      message.setSentDate(new Date());
      return message;
    } catch (Throwable t) {
      t.printStackTrace();
      return null;
    }
  }
}